### PR TITLE
Creating a standard starfish.wdl that can be run with any recipe file 

### PR DESCRIPTION
**Pull request by shanaxel42**
Created a standard starfish.wdl file that can be used to process per FOV. It takes in as inputs 
-path to experiment.json
-The number of fovs in the experiment
-the url path to a python recipe file

I think the easiest way to have user defined pipelines is to just allow them to upload a recipe file to their own repos, this way they don't need to make a PR to starfish to get a recipe in. Any recipe file just needs to implement a method called `process_fov(field_num: int, experiement_str: str)` that processes a single fov and returns a decoded intensity table. The starfish.wdl parallelizes per fov, downloads whatever user defined recipe file as recipe.py, imports it as a module called recipe and runs `recipe.process_fov(field_num: int, experiement_str: str)` it then saves each decoded intensity table and concatenates them all into decoded_concatenated.csv.

There's also three examples for the ISS spaceTX, ISS published, and merfish published pipelines. 

I also deleted the outdated old wdl directory. 

fixes: #1280 but probs needs some more documentation.

### CODE REVIEW

File: workflows/wdl/merfish_published/recipe.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+def process_fov(field_num: int, experiement_str: str):
+    fov_str: str = f"fov_{int(field_num):03d}"
+    # load experiment
+    experiment = starfish.Experiment.from_json(experiement_str)
+
+    print(f"loading fov: {fov_str}")
+    fov = experiment[fov_str]
+    imgs = fov.get_image(FieldOfView.PRIMARY_IMAGES)
+
+    print("gaussian high pass")
```
Comment (ambrosejcarr) [inline workflows/wdl/merfish_published/recipe.py]:
```suggestion
    print("Gaussian High Pass")
```

File: workflows/wdl/merfish_published/recipe.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    ghp = Filter.GaussianHighPass(sigma=3)
+    high_passed = ghp.run(imgs, verbose=True, in_place=False)
+
+    print("deconvoling")
```
Comment (ambrosejcarr) [inline workflows/wdl/merfish_published/recipe.py]:
```suggestion
    print("Deconvolve")
```

File: workflows/wdl/merfish_published/recipe.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    dpsf = Filter.DeconvolvePSF(num_iter=15, sigma=2, clip_method=Clip.SCALE_BY_CHUNK)
+    deconvolved = dpsf.run(high_passed, verbose=True, in_place=False)
+
+    print("guassian low pass")
```
Comment (ambrosejcarr) [inline workflows/wdl/merfish_published/recipe.py]:
```suggestion
    print("Gaussian Low Pass")
```

File: workflows/wdl/merfish_published/recipe.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        for t in experiment.extras['scale_factors']
+    }
+    filtered_imgs = deepcopy(low_passed)
+
+    for selector in imgs._iter_axes():
+        data = filtered_imgs.get_slice(selector)[0]
+        scaled = data / scale_factors[selector[Axes.ROUND.value], selector[Axes.CH.value]]
+        filtered_imgs.set_slice(selector, scaled, [Axes.ZPLANE])
+
+    print("decoding")
```
Comment (ambrosejcarr) [inline workflows/wdl/merfish_published/recipe.py]:
```suggestion
    print("Decode")
```

File: workflows/wdl/starfish.wdl [hunk 5]
```diff
@@ -0,0 +1,100 @@
+
+task process_field_of_view {
+
+    String experiment
+    Int field_of_view
+    String recipe_file
```
Comment (ambrosejcarr) [inline workflows/wdl/starfish.wdl]:
```suggestion
    File recipe_file
```

If you declare this a file, the WDL runner will localize it for you, then you can delete the wget line below.

File: workflows/wdl/starfish.wdl [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    command <<<
+        wget -O recipe.py ${recipe_file}
```
Comment (ambrosejcarr) [inline workflows/wdl/starfish.wdl]:
```suggestion
```

File: workflows/wdl/starfish.wdl (hunk 5 above)
Comment (shanaxel42) [reply workflows/wdl/starfish.wdl]: This does not work. The file has to be local already https://github.com/openwdl/wdl/blob/master/versions/1.0/SPEC.md#task-inputs

File: workflows/wdl/iss_published/recipe.py [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    fov = experiment[fov_str]
+
+    # note the structure of the 5D tensor containing the raw imaging data
+    imgs = fov.get_image(FieldOfView.PRIMARY_IMAGES)
+    dots = fov.get_image("dots")
+    nuclei = fov.get_image("nuclei")
+
+    masking_radius = 15
+    print("Filter WhiteTophat")
+    filt = Filter.WhiteTophat(masking_radius, is_volume=False)
```
Comment (ttung) [inline workflows/wdl/iss_published/recipe.py]: I'd just do:

```suggestion
    filt = Filter.WhiteTophat(masking_radius=15, is_volume=False)
```

File: workflows/wdl/iss_published/recipe.py [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+    filtered_imgs = filt.run(imgs, verbose=True, in_place=False)
```
Comment (ttung) [inline workflows/wdl/iss_published/recipe.py]: Consider moving the registration to before filtering so you can do inplace processing.

File: workflows/wdl/iss_published/recipe.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    print("Learning Transform")
+    learn_translation = LearnTransform.Translation(reference_stack=dots, axes=Axes.ROUND, upsampling=1000)
+    transforms_list = learn_translation.run(imgs.max_proj(Axes.CH, Axes.ZPLANE))
+
+    print("Applying transform")
+    warp = ApplyTransform.Warp()
+    registered_imgs = warp.run(filtered_imgs, transforms_list=transforms_list, in_place=False, verbose=True)
+
+    print("Detecting")
+    p = DetectSpots.BlobDetector(
```
Comment (ttung) [inline workflows/wdl/iss_published/recipe.py]:
```suggestion
    detector = DetectSpots.BlobDetector(
```

File: workflows/wdl/iss_published/recipe.py [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        num_sigma=30,
+        threshold=0.01,
+        measurement_type='mean',
+    )
+
+    intensities = p.run(registered_imgs, blobs_image=dots, blobs_axes=(Axes.ROUND, Axes.ZPLANE))
+
+    decoded = experiment.codebook.decode_per_round_max(intensities)
+    df = decoded.to_decoded_spots()
+    return df
```
Comment (ttung) [inline workflows/wdl/iss_published/recipe.py]: add nl.

File: workflows/wdl/iss_spaceTX/recipe.py [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    dots = imgs.max_proj(Axes.CH)
+
+    # filter
+    filt = starfish.image.Filter.WhiteTophat(masking_radius=15, is_volume=False)
+    filtered_imgs = filt.run(imgs, verbose=True, in_place=False)
+    filt.run(dots, verbose=True, in_place=True)
+
+    # find threshold
+    tmp = dots.sel({Axes.ROUND:0, Axes.CH:0, Axes.ZPLANE:0})
+    dots_threshold = np.percentile(np.ravel(tmp.xarray.values), 50)
```
Comment (ttung) [inline workflows/wdl/iss_spaceTX/recipe.py]: this is incompatible with the pipeline architecture.  do we have an issue for fixing this?

File: workflows/wdl/merfish_published/recipe.py [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+
+    print("Deconvolve")
+    dpsf = Filter.DeconvolvePSF(num_iter=15, sigma=2, clip_method=Clip.SCALE_BY_CHUNK)
+    deconvolved = dpsf.run(high_passed, verbose=True, in_place=False)
+
+    print("Guassian Low Pass")
+    glp = Filter.GaussianLowPass(sigma=1)
+    low_passed = glp.run(deconvolved, in_place=False, verbose=True)
+
+    scale_factors = {
```
Comment (ttung) [inline workflows/wdl/merfish_published/recipe.py]: this is also incompatible with the pipeline infrastructure.

File: workflows/wdl/merfish_published/recipe.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+        magnitude_threshold=1.77e-5,  # discard any pixel vectors below this magnitude
+        min_area=2,  # do not call a 'spot' if it's area is below this threshold (measured in pixels)
+        max_area=np.inf,  # do not call a 'spot' if it's area is above this threshold (measured in pixels)
+    )
+
+    initial_spot_intensities, prop_results = psd.run(filtered_imgs)
+
+    spot_intensities = initial_spot_intensities.loc[initial_spot_intensities[Features.PASSES_THRESHOLDS]]
+    df = spot_intensities.to_decoded_spots()
+    return df
```
Comment (ttung) [inline workflows/wdl/merfish_published/recipe.py]: nl after this.

File: workflows/wdl/starfish.wdl [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+    }
+}
+
+
+task concatenate_fovs {
+    Array[File] decoded_csvs
+
+    command <<<
+        python <<CODE
+        files = "${sep=' ' decoded_csvs}".strip().split()
```
Comment (ttung) [inline workflows/wdl/starfish.wdl]: it's kind of concerning that this is a bunch of python code floating in the ether.

Comment (shanaxel42) [reply workflows/wdl/starfish.wdl]: Yea I thought about moving this in starfish and making a wdl_utils file? But I thought that might be kind of random if it was the only wdl related thing...down to do that though if you think it makes more sense

**Pull request merged**